src/planner/partialize.c

```diff
@@ -185,6 +197,652 @@ partialize_agg_paths(RelOptInfo *rel)
+	subpath->pathtarget->sortgrouprefs = parent_path->pathtarget->sortgrouprefs;
+
+	if (can_sort)
+	{
+		*sorted_paths = lappend(*sorted_paths,
+								create_sorted_partially_agg_path(root,
+																 subpath,
+																 chunktarget,
+																 d_num_groups,
+																 extra_data));
+	}
+
+	if (can_hash)
+	{
+		*hashed_paths = lappend(*hashed_paths,
+								create_hashed_partially_agg_path(root,
+																 subpath,
+																 chunktarget,
+																 d_num_groups,
+																 extra_data));
+	}
+}
+
+/*
+ * Generate a total aggregation path for partial aggregations
+ */
+static void
+generate_agg_pushdown_path(PlannerInfo *root, Path *cheapest_total_path, RelOptInfo *output_rel,
+						   RelOptInfo *partially_grouped_rel, PathTarget *grouping_target,
+						   PathTarget *partial_grouping_target, bool can_sort, bool can_hash,
+						   double d_num_groups, GroupPathExtraData *extra_data)
+{
+	/* Get subpaths */
+	List *subpaths = get_subpaths_from_append_path(cheapest_total_path, false);
+
+	/* No subpaths available or unsupported append node */
+	if (subpaths == NIL)
+		return;
+
+	/* Generate agg paths on top of the append children */
+	List *sorted_subpaths = NIL;
+	List *hashed_subpaths = NIL;
+
+	ListCell *lc;
+	foreach (lc, subpaths)
+	{
+		Path *subpath = lfirst(lc);
+
+		/* Check if we have an append path under an append path (e.g., a partially compressed
+		 * chunk)*/
+		List *subsubpaths = get_subpaths_from_append_path(subpath, false);
+
+		if (subsubpaths != NIL)
+		{
+			List *sorted_subsubpaths = NIL;
+			List *hashed_subsubpaths = NIL;
+
+			ListCell *lc2;
+			foreach (lc2, subsubpaths)
+			{
+				Path *subsubpath = lfirst(lc2);
+
+				add_partially_aggregated_subpaths(root,
+												  cheapest_total_path,
+												  partial_grouping_target,
+												  d_num_groups,
+												  extra_data,
+												  can_sort,
+												  can_hash,
+												  subsubpath,
+												  &sorted_subsubpaths /* Result path */,
+												  &hashed_subsubpaths /* Result path */);
+			}
+
+			if (can_sort)
+			{
+				sorted_subpaths = lappend(sorted_subpaths,
+										  copy_append_like_path(root, subpath, sorted_subsubpaths));
+			}
+
+			if (can_hash)
+			{
+				hashed_subpaths = lappend(hashed_subpaths,
+										  copy_append_like_path(root, subpath, hashed_subsubpaths));
+			}
+		}
+		else
+		{
+			add_partially_aggregated_subpaths(root,
+											  cheapest_total_path,
+											  partial_grouping_target,
+											  d_num_groups,
+											  extra_data,
+											  can_sort,
+											  can_hash,
+											  subpath,
+											  &sorted_subpaths /* Result paths */,
+											  &hashed_subpaths /* Result paths */);
+		}
+	}
+
+	/* Create new append paths */
+	cheapest_total_path->pathtarget = partial_grouping_target;
+
+	if (sorted_subpaths != NIL)
+	{
+		add_path(partially_grouped_rel,
```

Comment:
Trying to understand when to use `add_path` vs `add_partial_path` and why we aren't using the latter.

Is it because `add_partial_path` is only for parallel-executed paths and not "push-down" paths? (In other words, is `add_partial_path`  mislabeled and should be called something like `add_parallel_path` or `add_parallel_partial_path`?)

Reply:
It is hard to from the naming to distinguish between partial aggregates and partial paths. So, I guess `add_parallel_path` would be a better name. However, it is a PostgreSQL function.

I use `add_partial_path` to store all partial (parallel) paths that need to be finished later by a gather node (see Line 577 - `foreach (lc, partially_grouped_rel->partial_pathlist)`). A gather node is created and the resulting path is added calling `add_path`.

Reply:
Not sure I follow. Doesn't all partial paths require finalization with a Gather node, regardless of whether they are parallel or not?

I am just trying to understand whether we are adding paths to the right path "list" as there are two "pathlists"  in a RelOptInfo. One `rel->pathlist` and one `rel->partial_pathlist`. I don't recall the difference between these two so I am asking to ensure we are doing the right thing here.